In [1]:
#import re

#texto = input("Ingrese un texto: ")
#palabras = re.findall(r"\b\w+\b", texto)
#conteo = len(palabras)



#print(f"El texto tiene {conteo} palabras.")


In [2]:
#!pip install streamlit

In [3]:
import numpy as np
import pandas as pd
from pathlib import Path

# Crear nuevamente un archivo Excel amplio y reproducible
archivo = Path('datos_ejemplo.xlsx')
rng = np.random.default_rng(42)
registros = 1000

productos = [
    'Laptop Pro 14', 'Monitor 27', 'Teclado mecanico', 'Mouse inalambrico',
    'Auriculares Bluetooth', 'Tablet Plus', 'Camara web HD', 'Disco SSD 1TB',
    'Impresora multifuncion', 'Silla ergonomica', 'Router WiFi 6', 'Smartwatch'
]
categorias = ['Tecnologia', 'Accesorios', 'Oficina', 'Movilidad']
regiones = ['Norte', 'Sur', 'Este', 'Oeste', 'Centro']
metodos_pago = ['Tarjeta', 'Efectivo', 'Transferencia', 'PayPal']

fecha = pd.Timestamp('2024-01-01') + pd.to_timedelta(rng.integers(0, 366, registros), unit='D')
producto = rng.choice(productos, registros)
categoria = rng.choice(categorias, registros, p=[0.35, 0.30, 0.20, 0.15])
region = rng.choice(regiones, registros)
unidades = rng.integers(1, 12, registros)
clientes = rng.integers(1, 30, registros)
precio_unitario = rng.uniform(15, 650, registros).round(2)
ventas = (unidades * precio_unitario).round(2)
descuentos = (ventas * rng.uniform(0.00, 0.18, registros)).round(2)
costo = (ventas * rng.uniform(0.42, 0.78, registros)).round(2)
ganancia = (ventas - descuentos - costo).round(2)
margen = (ganancia / ventas).round(4)
ticket_promedio = (ventas / clientes).round(2)

df = pd.DataFrame({
    'Fecha': fecha,
    'Producto': producto,
    'Ventas': ventas,
    'Unidades': unidades,
    'Categoria': categoria,
    'Region': region,
    'Clientes': clientes,
    'Metodo_Pago': rng.choice(metodos_pago, registros),
    'Costo': costo,
    'Ganancia': ganancia,
    'Margen': margen,
    'Descuentos': descuentos,
    'Ticket_Promedio': ticket_promedio,
}).sort_values('Fecha').reset_index(drop=True)

df.to_excel(archivo, index=False)
print(f'Archivo recreado: {archivo.resolve()}')
print(f'Registros creados: {len(df)}')
print(f'Productos distintos: {df["Producto"].nunique()}')
print(df.head())
print(f'Total de ventas: {df["Ventas"].sum():,.2f}')
print(f'Promedio de ventas: {df["Ventas"].mean():,.2f}')


Archivo recreado: D:\alejandro_palencia\Curso Python Daxus\Proyecto04\datos_ejemplo.xlsx
Registros creados: 1000
Productos distintos: 12
       Fecha               Producto   Ventas  Unidades   Categoria  Region  \
0 2024-01-02             Smartwatch   941.20         2  Accesorios  Centro   
1 2024-01-02          Disco SSD 1TB  1236.04         2  Accesorios   Oeste   
2 2024-01-02          Camara web HD  5739.03        11  Accesorios     Sur   
3 2024-01-02  Auriculares Bluetooth  2637.00         5  Accesorios   Norte   
4 2024-01-03          Router WiFi 6   701.97         3     Oficina    Este   

   Clientes    Metodo_Pago    Costo  Ganancia  Margen  Descuentos  \
0         9  Transferencia   500.87    414.96  0.4409       25.37   
1        20       Efectivo   803.00    417.60  0.3379       15.44   
2         2         PayPal  2743.21   2291.49  0.3993      704.33   
3        28         PayPal  1671.17    574.13  0.2177      391.70   
4         4         PayPal   470.30    201.89  0.

# ETAPA 1 — CARGA Y REVISIÓN INICIAL

import pandas as pd
import numpy as np

# Cargar el archivo Excel.
# Si no existe, se genera un dataset de ejemplo mínimo con la estructura pedida para continuar el análisis.
archivo = 'datos_ejemplo.xlsx'

if not pd.io.common.file_exists(archivo):
    df = pd.DataFrame({
        'Fecha': pd.date_range('2024-01-01', periods=5, freq='D'),
        'Ventas': [1200, 1500, 980, 1800, 1400],
        'Unidades': [20, 25, 18, 30, 22],
        'Categoria': ['A', 'B', 'A', 'C', 'B'],
        'Region': ['Norte', 'Sur', 'Este', 'Oeste', 'Norte'],
        'Clientes': [50, 60, 45, 70, 55],
        'Metodo_Pago': ['Tarjeta', 'Efectivo', 'Tarjeta', 'Transferencia', 'Tarjeta'],
        'Costo': [700, 800, 600, 900, 750],
        'Ganancia': [500, 700, 380, 900, 650],
        'Margen': [0.42, 0.47, 0.39, 0.50, 0.46],
        'Descuentos': [50, 100, 30, 80, 60],
        'Ticket_Promedio': [24, 25, 22, 26, 25]
    })
    df.to_excel(archivo, index=False)
    print(f'Archivo no encontrado; se creó un dataset de ejemplo en: {archivo}')
else:
    df = pd.read_excel(archivo)

print('=== REVISIÓN INICIAL ===')
print(f'Dimensiones del dataset: {df.shape}')
print('\nNombres de columnas:')
print(df.columns.tolist())
print('\nTipos de datos:')
print(df.dtypes)
print('\nValores nulos por columna:')
print(df.isnull().sum())
print('\nDuplicados:')
print(f'Número de filas duplicadas: {df.duplicated().sum()}')

# Valores únicos de variables categóricas
print('\nValores únicos por variable categórica:')
for col in df.select_dtypes(include=['object', 'category']).columns:
    print(f'- {col}: {df[col].dropna().unique().tolist()}')

print('\nEstadísticas descriptivas:')
print(df.describe(include='all').T)

print('\nVerificación de columnas numéricas clave:')
for col in ['Ventas', 'Unidades', 'Costo', 'Ganancia', 'Descuentos', 'Ticket_Promedio']:
    if col in df.columns:
        negativos = (pd.to_numeric(df[col], errors='coerce') < 0).sum()
        no_numericos = pd.to_numeric(df[col], errors='coerce').isna().sum()
        print(f'- {col}: negativos={negativos}, no_numéricos={no_numericos}')


In [4]:
# ETAPA 1 — CARGA Y REVISIÓN INICIAL

## Objetivo
# Se valida la estructura real del archivo Excel antes de analizarlo.

import pandas as pd
import numpy as np
from pathlib import Path

archivo = Path('datos_ejemplo.xlsx')

if not archivo.exists():
    raise FileNotFoundError(
        f'No se encontró el archivo {archivo}. Colócalo en la misma carpeta del notebook antes de ejecutar el análisis.'
    )

# Carga del archivo
ventas = pd.read_excel(archivo)

print('=== REVISIÓN INICIAL ===')
print(f'Dimensiones del dataset: {ventas.shape}')
print('\nColumnas:')
print(list(ventas.columns))
print('\nTipos de datos:')
print(ventas.dtypes)
print('\nValores nulos:')
print(ventas.isnull().sum())
print('\nDuplicados:')
print(f'Filas duplicadas: {ventas.duplicated().sum()}')
print('\nValores únicos por columnas categóricas:')
for col in ventas.select_dtypes(include=['object', 'category']).columns:
    print(f'- {col}: {ventas[col].dropna().unique()[:10].tolist()}')
print('\nEstadísticas descriptivas:')
print(ventas.describe(include='all').T)


=== REVISIÓN INICIAL ===
Dimensiones del dataset: (1000, 13)

Columnas:
['Fecha', 'Producto', 'Ventas', 'Unidades', 'Categoria', 'Region', 'Clientes', 'Metodo_Pago', 'Costo', 'Ganancia', 'Margen', 'Descuentos', 'Ticket_Promedio']

Tipos de datos:
Fecha              datetime64[us]
Producto                      str
Ventas                    float64
Unidades                    int64
Categoria                     str
Region                        str
Clientes                    int64
Metodo_Pago                   str
Costo                     float64
Ganancia                  float64
Margen                    float64
Descuentos                float64
Ticket_Promedio           float64
dtype: object

Valores nulos:
Fecha              0
Producto           0
Ventas             0
Unidades           0
Categoria          0
Region             0
Clientes           0
Metodo_Pago        0
Costo              0
Ganancia           0
Margen             0
Descuentos         0
Ticket_Promedio    0
dtype: i

C:\Users\alejandro_palencia\AppData\Local\Temp\ipykernel_25400\844182944.py:31: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  for col in ventas.select_dtypes(include=['object', 'category']).columns:


## ETAPA 2 — LIMPIEZA Y TRANSFORMACIÓN

En esta etapa se revisa el formato de cada variable y se crea una copia limpia para el análisis. La conversión de `Fecha` a datetime y la normalización de columnas numéricas es necesaria para poder calcular ventas, márgenes y series temporales sin errores de tipo.


In [5]:
# ETAPA 2 — LIMPIEZA Y TRANSFORMACIÓN

import pandas as pd
import numpy as np

# Copia limpia del dataset para no alterar los datos originales
ventas_clean = ventas.copy()

# 1) Fecha: convertir a datetime
# Esto permite trabajar con series temporales y agrupar por mes o día correctamente.
ventas_clean['Fecha'] = pd.to_datetime(ventas_clean['Fecha'])

# 2) Variables numéricas: forzar tipos adecuados
# Se corrigen columnas que deben ser numéricas para evitar errores en sumas, promedios y comparaciones.
for col in ['Ventas', 'Unidades', 'Clientes', 'Costo', 'Ganancia', 'Descuentos', 'Ticket_Promedio']:
    ventas_clean[col] = pd.to_numeric(ventas_clean[col], errors='coerce')

# 3) Margen: asegurar que esté en rango lógico
# Un margen fuera de 0 a 1 puede indicar dato inconsistente.
ventas_clean['Margen'] = pd.to_numeric(ventas_clean['Margen'], errors='coerce')
ventas_clean['Margen'] = ventas_clean['Margen'].clip(lower=0, upper=1)

# 4) Revisión de nulos y duplicados
print('Valores nulos tras la limpieza:')
print(ventas_clean.isnull().sum())
print('\nDuplicados tras la limpieza:')
print(ventas_clean.duplicated().sum())

# 5) Revisar valores negativos en columnas clave
print('\nValores negativos en columnas clave:')
for col in ['Ventas', 'Unidades', 'Costo', 'Ganancia', 'Descuentos', 'Ticket_Promedio']:
    print(f'- {col}: {(ventas_clean[col] < 0).sum()}')

# 6) Ver datos finales limpios
print('\nVista general del dataset limpio:')
print(ventas_clean.head())


Valores nulos tras la limpieza:
Fecha              0
Producto           0
Ventas             0
Unidades           0
Categoria          0
Region             0
Clientes           0
Metodo_Pago        0
Costo              0
Ganancia           0
Margen             0
Descuentos         0
Ticket_Promedio    0
dtype: int64

Duplicados tras la limpieza:
0

Valores negativos en columnas clave:
- Ventas: 0
- Unidades: 0
- Costo: 0
- Ganancia: 0
- Descuentos: 0
- Ticket_Promedio: 0

Vista general del dataset limpio:
       Fecha               Producto   Ventas  Unidades   Categoria  Region  \
0 2024-01-02             Smartwatch   941.20         2  Accesorios  Centro   
1 2024-01-02          Disco SSD 1TB  1236.04         2  Accesorios   Oeste   
2 2024-01-02          Camara web HD  5739.03        11  Accesorios     Sur   
3 2024-01-02  Auriculares Bluetooth  2637.00         5  Accesorios   Norte   
4 2024-01-03          Router WiFi 6   701.97         3     Oficina    Este   

   Clientes    Metod

# ETAPA 3 — DOCUMENTACIÓN DEL EDA

## Análisis Exploratorio de Datos — Ventas 2024

### 1. Objetivo del análisis
Este notebook tiene como objetivo revisar la calidad de los datos, limpiar la estructura, analizar el comportamiento de las ventas y evaluar la rentabilidad del negocio para 2024.

### 2. Descripción del dataset
Se trabaja con un archivo Excel de ventas diarias con columnas como Fecha, Ventas, Unidades, Categoria, Region, Clientes, Metodo_Pago, Costo, Ganancia, Margen, Descuentos y Ticket_Promedio.

### 3. Calidad y estructura de los datos
Se verificó dimensiones, tipos, nulos, duplicados y variables categóricas. La estructura inicial del archivo fue consistente y no se detectaron valores faltantes ni duplicados.

### 4. Limpieza y transformación
Se convirtió la columna Fecha a tipo datetime y se normalizaron las columnas numéricas para garantizar cálculos fiables. El margen se limitó a un rango lógico entre 0 y 1.

### 5. Análisis descriptivo
Se utilizarán métricas como total de ventas, promedio diario, unidades, costo, ganancia, margen y ticket promedio para resumir el comportamiento del negocio.

### 6. Análisis de ventas
Se revisará la evolución de ventas por periodo, el volumen por categoría y región, y el comportamiento del ticket promedio.

### 7. Análisis de rentabilidad
Se evaluará la relación entre ventas, costos, ganancias y descuentos para detectar segmentos con mejor o peor rentabilidad.

### 8. Análisis por categoría y región
Se compararán categorías y regiones según volumen, margen y contribución a la rentabilidad.

### 9. Visualizaciones
Se generarán gráficos interactivos con Plotly para resumir tendencias y comparaciones relevantes.

### 10. Hallazgos principales
Se identificarán los principales patrones observados en las ventas, costos, rentabilidad y descuentos.

### 11. Conclusiones
Las conclusiones se basarán únicamente en los datos cargados y validados.

### 12. Recomendaciones
Se propondrán recomendaciones de negocio basadas en rentabilidad, mix de categorías y gestión de descuentos.


In [6]:
# ETAPA 4 — ANÁLISIS GENERAL

# Resumen financiero y operativo
ventas_summary = {
    'Ventas_totales': ventas_clean['Ventas'].sum(),
    'Ventas_promedio_diarias': ventas_clean['Ventas'].mean(),
    'Ventas_mediana': ventas_clean['Ventas'].median(),
    'Registros': len(ventas_clean),
    'Unidades_vendidas': ventas_clean['Unidades'].sum(),
    'Costo_total': ventas_clean['Costo'].sum(),
    'Ganancia_total': ventas_clean['Ganancia'].sum(),
    'Margen_promedio': ventas_clean['Margen'].mean(),
    'Margen_ponderado': ventas_clean['Ganancia'].sum() / ventas_clean['Ventas'].sum(),
    'Descuentos_totales': ventas_clean['Descuentos'].sum(),
    'Ticket_promedio': ventas_clean['Ticket_Promedio'].mean(),
    'Clientes_totales': ventas_clean['Clientes'].sum(),
}

print('=== RESUMEN GENERAL ===')
for k, v in ventas_summary.items():
    print(f'{k}: {v}')

# Evolución mensual
ventas_clean['Mes'] = ventas_clean['Fecha'].dt.to_period('M').astype(str)
resumen_mensual = ventas_clean.groupby('Mes').agg(
    Ventas=('Ventas', 'sum'),
    Ganancia=('Ganancia', 'sum'),
    Costo=('Costo', 'sum'),
    Unidades=('Unidades', 'sum'),
    Ticket_Promedio=('Ticket_Promedio', 'mean'),
    Descuentos=('Descuentos', 'sum')
).reset_index()
resumen_mensual['Margen'] = resumen_mensual['Ganancia'] / resumen_mensual['Ventas']

print('\n=== RESUMEN MENSUAL ===')
print(resumen_mensual)

# Rentabilidad por categoría
resumen_categoria = ventas_clean.groupby('Categoria').agg(
    Ventas=('Ventas', 'sum'),
    Ganancia=('Ganancia', 'sum'),
    Unidades=('Unidades', 'sum'),
    Clientes=('Clientes', 'sum'),
    Descuentos=('Descuentos', 'sum')
).reset_index()
resumen_categoria['Margen'] = resumen_categoria['Ganancia'] / resumen_categoria['Ventas']
print('\n=== RESUMEN POR CATEGORÍA ===')
print(resumen_categoria)

# Rentabilidad por región
resumen_region = ventas_clean.groupby('Region').agg(
    Ventas=('Ventas', 'sum'),
    Ganancia=('Ganancia', 'sum'),
    Unidades=('Unidades', 'sum'),
    Clientes=('Clientes', 'sum'),
    Descuentos=('Descuentos', 'sum')
).reset_index()
resumen_region['Margen'] = resumen_region['Ganancia'] / resumen_region['Ventas']
print('\n=== RESUMEN POR REGIÓN ===')
print(resumen_region)


=== RESUMEN GENERAL ===
Ventas_totales: 1977619.19
Ventas_promedio_diarias: 1977.61919
Ventas_mediana: 1549.12
Registros: 1000
Unidades_vendidas: 6018
Costo_total: 1185719.3199999998
Ganancia_total: 622034.09
Margen_promedio: 0.3156444
Margen_ponderado: 0.31453683962279916
Descuentos_totales: 169865.78
Ticket_promedio: 282.75349
Clientes_totales: 14260

=== RESUMEN MENSUAL ===
        Mes     Ventas  Ganancia      Costo  Unidades  Ticket_Promedio  \
0   2024-01  153223.15  50276.15   90312.99       423       321.819859   
1   2024-02  201238.37  62636.12  121808.23       670       222.636667   
2   2024-03  126810.72  39394.93   77384.66       409       268.442267   
3   2024-04  150338.18  47125.45   90134.29       501       243.189878   
4   2024-05  104582.81  36100.59   61722.67       324       311.880702   
5   2024-06  188459.49  54065.45  116920.39       616       230.952062   
6   2024-07  189201.57  62370.74  111631.39       506       354.492088   
7   2024-08  175549.11  5622

In [7]:
# ETAPA 5 — VISUALIZACIONES INTERACTIVAS

import plotly.express as px

# Gráfico 1: evolución mensual de ventas y ganancia
fig_ventas = px.line(
    resumen_mensual,
    x='Mes',
    y=['Ventas', 'Ganancia'],
    title='Evolución mensual de ventas y ganancia',
    labels={'value': 'Monto', 'variable': 'Indicador', 'Mes': 'Mes'}
)
fig_ventas.show()

# Gráfico 2: ventas y ganancia por categoría
fig_cat = px.bar(
    resumen_categoria,
    x='Categoria',
    y=['Ventas', 'Ganancia'],
    barmode='group',
    title='Ventas y ganancia por categoría',
    labels={'value': 'Monto', 'variable': 'Indicador'}
)
fig_cat.show()

# Gráfico 3: margen por categoría
fig_margin_cat = px.bar(
    resumen_categoria,
    x='Categoria',
    y='Margen',
    title='Rentabilidad por categoría (margen)',
    labels={'Margen': 'Margen', 'Categoria': 'Categoría'}
)
fig_margin_cat.show()

# Gráfico 4: ventas y margen por región
fig_region = px.bar(
    resumen_region,
    x='Region',
    y=['Ventas', 'Ganancia'],
    barmode='group',
    title='Ventas y ganancia por región',
    labels={'value': 'Monto', 'variable': 'Indicador'}
)
fig_region.show()

# Gráfico 5: relación entre ventas, costo y ganancia
fig_rel = px.scatter(
    ventas_clean,
    x='Ventas',
    y='Ganancia',
    size='Costo',
    color='Categoria',
    title='Relación entre ventas, costo y ganancia',
    labels={'Ventas': 'Ventas', 'Ganancia': 'Ganancia', 'Costo': 'Costo'}
)
fig_rel.show()

# Gráfico 6: impacto de descuentos sobre el margen
fig_desc = px.scatter(
    ventas_clean,
    x='Descuentos',
    y='Margen',
    color='Categoria',
    size='Ventas',
    title='Impacto de los descuentos sobre el margen',
    labels={'Descuentos': 'Descuentos', 'Margen': 'Margen', 'Ventas': 'Ventas'}
)
fig_desc.show()

# Gráfico 7: ticket promedio por categoría
fig_ticket = px.bar(
    ventas_clean.groupby('Categoria')['Ticket_Promedio'].mean().reset_index(),
    x='Categoria',
    y='Ticket_Promedio',
    title='Ticket promedio por categoría',
    labels={'Ticket_Promedio': 'Ticket promedio', 'Categoria': 'Categoría'}
)
fig_ticket.show()


In [8]:
# ETAPA 6 — TABLAS DE ANÁLISIS

# Resumen mensual
print('=== RESUMEN MENSUAL ===')
print(resumen_mensual.round(2).to_string(index=False))

# Ranking por categoría
ranking_categoria = resumen_categoria.sort_values('Ventas', ascending=False).reset_index(drop=True)
print('\n=== RANKING POR CATEGORÍA ===')
print(ranking_categoria.round(2).to_string(index=False))

# Ranking por región
ranking_region = resumen_region.sort_values('Ventas', ascending=False).reset_index(drop=True)
print('\n=== RANKING POR REGIÓN ===')
print(ranking_region.round(2).to_string(index=False))

# Tabla de rentabilidad por categoría con margen y participación
categoria_rentabilidad = resumen_categoria[['Categoria', 'Ventas', 'Ganancia', 'Margen']].copy()
categoria_rentabilidad['Participacion_Ventas'] = categoria_rentabilidad['Ventas'] / categoria_rentabilidad['Ventas'].sum()
print('\n=== RENTABILIDAD POR CATEGORÍA ===')
print(categoria_rentabilidad.round(4).to_string(index=False))


=== RESUMEN MENSUAL ===
    Mes    Ventas  Ganancia     Costo  Unidades  Ticket_Promedio  Descuentos  Margen
2024-01 153223.15  50276.15  90312.99       423           321.82    12634.01    0.33
2024-02 201238.37  62636.12 121808.23       670           222.64    16794.02    0.31
2024-03 126810.72  39394.93  77384.66       409           268.44    10031.13    0.31
2024-04 150338.18  47125.45  90134.29       501           243.19    13078.44    0.31
2024-05 104582.81  36100.59  61722.67       324           311.88     6759.55    0.35
2024-06 188459.49  54065.45 116920.39       616           230.95    17473.65    0.29
2024-07 189201.57  62370.74 111631.39       506           354.49    15199.44    0.33
2024-08 175549.11  56223.73 104574.56       534           351.97    14750.82    0.32
2024-09 160504.31  53566.76  92364.90       441           271.43    14572.65    0.33
2024-10 205842.05  63615.52 123517.80       607           312.72    18708.73    0.31
2024-11 128205.89  38363.80  77274.61    

In [9]:
# ETAPA 7 — HALLAZGOS DE NEGOCIO

# Hallazgos basados en los datos reales del dataset
print('=== HALLAZGOS DE NEGOCIO ===')
print(f'- Mayor volumen de ventas: Categoria {resumen_categoria.loc[resumen_categoria["Ventas"].idxmax(), "Categoria"]} con {resumen_categoria["Ventas"].max()} en ventas.')
print(f'- Mayor margen: Categoria {resumen_categoria.loc[resumen_categoria["Margen"].idxmax(), "Categoria"]} con margen {resumen_categoria["Margen"].max():.2%}.')
print(f'- Región con mayor volumen: {resumen_region.loc[resumen_region["Ventas"].idxmax(), "Region"]} con {resumen_region["Ventas"].max()} en ventas.')
print(f'- Región con mayor margen: {resumen_region.loc[resumen_region["Margen"].idxmax(), "Region"]} con margen {resumen_region["Margen"].max():.2%}.')
print(f'- Total de ganancia: {ventas_clean["Ganancia"].sum()}')
print(f'- Margen ponderado: {ventas_clean["Ganancia"].sum() / ventas_clean["Ventas"].sum():.2%}')
print(f'- Descuentos totales: {ventas_clean["Descuentos"].sum()}')
print(f'- Ticket promedio: {ventas_clean["Ticket_Promedio"].mean():.2f}')

# Interpretación de rentabilidad por categoría y región
print('\nInterpretación:')
print(f'- La categoría {resumen_categoria.loc[resumen_categoria["Ventas"].idxmax(), "Categoria"]} concentra más volumen de ventas y también aporta la mayor ganancia absoluta.')
print(f'- La categoría {resumen_categoria.loc[resumen_categoria["Margen"].idxmax(), "Categoria"]} tiene el mayor margen, con {resumen_categoria["Margen"].max():.2%}.')
print(f'- La región {resumen_region.loc[resumen_region["Ventas"].idxmax(), "Region"]} concentra la mayor parte de las ventas; la región {resumen_region.loc[resumen_region["Margen"].idxmax(), "Region"]} destaca por mayor margen.')
print('- Los descuentos tienen impacto directo sobre la rentabilidad y deben revisarse si aumentan mucho sin mejorar volumen.')


=== HALLAZGOS DE NEGOCIO ===
- Mayor volumen de ventas: Categoria Tecnologia con 703865.57 en ventas.
- Mayor margen: Categoria Tecnologia con margen 32.22%.
- Región con mayor volumen: Sur con 427374.09 en ventas.
- Región con mayor margen: Norte con margen 32.42%.
- Total de ganancia: 622034.09
- Margen ponderado: 31.45%
- Descuentos totales: 169865.78
- Ticket promedio: 282.75

Interpretación:
- La categoría Tecnologia concentra más volumen de ventas y también aporta la mayor ganancia absoluta.
- La categoría Tecnologia tiene el mayor margen, con 32.22%.
- La región Sur concentra la mayor parte de las ventas; la región Norte destaca por mayor margen.
- Los descuentos tienen impacto directo sobre la rentabilidad y deben revisarse si aumentan mucho sin mejorar volumen.


# ETAPA 8 — INFORME GERENCIAL DE RENTABILIDAD

## Resumen ejecutivo
El dataset actual contiene 1,000 registros y muestra ventas totales de 1,977,619.19, una ganancia total de 622,034.09 y un margen ponderado de 31.45%. La categoría Tecnologia concentra el mayor volumen de ventas y también presenta el mayor margen. La región Sur lidera las ventas, mientras que la región Norte presenta el mayor margen.

## Indicadores clave
- Ventas: 1,977,619.19
- Ganancia: 622,034.09
- Margen ponderado: 31.45%
- Costos: 1,185,719.32
- Unidades: 6,018
- Clientes: 14,260
- Ticket promedio: 282.75
- Descuentos: 169,865.78

## Principales hallazgos
1. La categoría Tecnologia lidera el volumen de ventas y la ganancia absoluta.
2. La categoría Tecnologia presenta el mayor margen, con 32.22%.
3. La región Sur aporta el mayor volumen de ventas.
4. La región Norte presenta el mayor margen, con 32.42%.
5. El margen ponderado general se sitúa en 31.45%.
6. Los descuentos deben analizarse junto con su efecto sobre la ganancia y el margen.

## Rentabilidad
La categoría Tecnologia lidera tanto las ventas como el margen dentro del dataset actual. La región Sur concentra el mayor volumen, mientras que la región Norte destaca por su rentabilidad relativa. La estrategia debe equilibrar crecimiento comercial con protección del margen.

## Riesgos y oportunidades
- Riesgo: descuentos elevados sin mejora equivalente en volumen o ganancia.
- Oportunidad: fortalecer la categoría Tecnologia, que lidera ventas y margen.
- Oportunidad: analizar las prácticas de la región Norte para mejorar la rentabilidad en otras regiones.
- Oportunidad: mantener la cobertura comercial de la región Sur por su liderazgo en ventas.

## Conclusiones
El dataset actual presenta ventas por 1,977,619.19 y una ganancia de 622,034.09, con un margen ponderado de 31.45%. La mayor contribución comercial corresponde a la categoría Tecnologia y a la región Sur, mientras que la región Norte presenta el mayor margen. La gestión de costos, descuentos y mix de categorías es fundamental para sostener la rentabilidad.

## Recomendaciones
- Mantener y fortalecer la categoría Tecnologia, evaluando conjuntamente ventas, ganancia y margen.
- Conservar la presencia comercial en la región Sur por su liderazgo en ventas.
- Revisar las prácticas de la región Norte para identificar acciones que favorezcan un mayor margen.
- Establecer límites y seguimiento periódico para los descuentos.
- Actualizar el archivo Excel y ejecutar nuevamente el notebook para comparar periodos.

# ETAPA 9 — PREPARACIÓN PARA PDF

## Informe gerencial final para exportación

Este notebook está organizado en bloques temáticos para facilitar su reutilización mensual:
- carga y revisión inicial,
- limpieza y transformación,
- EDA y documentación,
- análisis comercial,
- visualizaciones interactivas,
- tablas resumen,
- hallazgos y recomendaciones,
- informe gerencial final.

### Requisitos para reutilización
Para actualizar el informe con nuevos datos, basta con:
1. reemplazar el archivo Excel,
2. ejecutar todas las celdas desde el inicio,
3. revisar que las tablas y gráficos se actualicen automáticamente.

### Gráficos clave para incluir en el PDF
- evolución mensual de ventas y ganancia,
- ventas y ganancia por categoría,
- margen por categoría,
- ventas y ganancia por región,
- relación entre ventas, costos y ganancia,
- impacto de descuentos sobre el margen.

### Observación final
Los resultados del presente análisis están basados exclusivamente en los datos del archivo cargado y validados en el notebook. Las conclusiones no se extrapolan fuera del conjunto de datos analizado.

## Resultado final
El negocio presenta un margen ponderado de 31.45%. La categoría Tecnologia lidera tanto las ventas como el margen, mientras que la región Sur concentra el mayor volumen de ventas y la región Norte presenta el mayor margen. El producto Camara web HD registra la mayor facturación acumulada. El control de descuentos y la optimización del mix de categorías y regiones son aspectos relevantes para sostener la rentabilidad en periodos futuros.


In [10]:
# ETAPA 10 — EXPORTAR INFORME A PDF Y HTML INTERACTIVO

# Instalar en el kernel ACTUAL: data (Python 3.13.15)
%pip install --upgrade kaleido reportlab matplotlib

from pathlib import Path
import matplotlib.pyplot as plt
from reportlab.lib.pagesizes import letter
from reportlab.platypus import SimpleDocTemplate, Paragraph, Spacer, Image
from reportlab.lib.styles import getSampleStyleSheet

# Directorio para guardar reportes
report_dir = Path('reportes')
report_dir.mkdir(parents=True, exist_ok=True)
print(f'Directorio de reportes: {report_dir.resolve()}')

# Exportar gráficas estáticas con Matplotlib para el PDF.
def guardar_grafica(nombre, titulo, x, series, ylabel='Monto'):
    figura, eje = plt.subplots(figsize=(12, 7))
    for etiqueta, valores, estilo in series:
        eje.plot(x, valores, marker='o', linewidth=2, label=etiqueta, color=estilo)
    eje.set_title(titulo)
    eje.set_xlabel('Categoría / periodo')
    eje.set_ylabel(ylabel)
    eje.grid(alpha=0.25)
    eje.legend()
    figura.tight_layout()
    figura.savefig(report_dir / nombre, dpi=150)
    plt.close(figura)
    print(f'Imagen creada: {report_dir / nombre}')

categorias = resumen_categoria['Categoria'].astype(str).tolist()
regiones = resumen_region['Region'].astype(str).tolist()
meses = resumen_mensual['Mes'].astype(str).tolist()

for nombre, titulo, eje_x, series, ylabel in [
    ('ventas_mensual.png', 'Evolución mensual de ventas y ganancia', meses, [
        ('Ventas', resumen_mensual['Ventas'], '#2563eb'),
        ('Ganancia', resumen_mensual['Ganancia'], '#dc2626'),
    ], 'Monto'),
    ('ventas_categoria.png', 'Ventas y ganancia por categoría', categorias, [
        ('Ventas', resumen_categoria['Ventas'], '#2563eb'),
        ('Ganancia', resumen_categoria['Ganancia'], '#dc2626'),
    ], 'Monto'),
    ('margen_categoria.png', 'Margen por categoría', categorias, [
        ('Margen', resumen_categoria['Margen'] * 100, '#059669'),
    ], 'Margen (%)'),
    ('ventas_region.png', 'Ventas y ganancia por región', regiones, [
        ('Ventas', resumen_region['Ventas'], '#2563eb'),
        ('Ganancia', resumen_region['Ganancia'], '#dc2626'),
    ], 'Monto'),
]:
    guardar_grafica(nombre, titulo, eje_x, series, ylabel)

# Hallazgos calculados para incorporarlos también al HTML.
producto_top = ventas_clean.groupby('Producto')['Ventas'].sum().idxmax()
categoria_top = resumen_categoria.loc[resumen_categoria['Ventas'].idxmax(), 'Categoria']
categoria_margen = resumen_categoria.loc[resumen_categoria['Margen'].idxmax(), 'Categoria']
region_top = resumen_region.loc[resumen_region['Ventas'].idxmax(), 'Region']
region_margen = resumen_region.loc[resumen_region['Margen'].idxmax(), 'Region']

# 1) Crear informe HTML interactivo con las gráficas Plotly y el análisis gerencial
figuras_interactivas = [fig_ventas, fig_cat, fig_margin_cat, fig_region, fig_rel, fig_desc, fig_ticket]
html_parts = [
    "<html><head><meta charset='utf-8'><title>Informe interactivo de ventas</title>",
    "<style>body { font-family: Arial, sans-serif; margin: 30px; line-height: 1.5; } h1, h2 { color: #1f2937; } section { margin-bottom: 28px; } li { margin-bottom: 8px; }</style>",
    "</head><body>",
    "<h1>Informe interactivo de rentabilidad</h1>",
    "<section><h2>Resumen ejecutivo</h2>",
    f"<p>Registros: {ventas_summary['Registros']:,}</p>",
    f"<p>Ventas totales: {ventas_summary['Ventas_totales']:,.2f}</p>",
    f"<p>Ganancia total: {ventas_summary['Ganancia_total']:,.2f}</p>",
    f"<p>Margen ponderado: {ventas_summary['Margen_ponderado']:.2%}</p>",
    f"<p>Unidades: {ventas_summary['Unidades_vendidas']:,}</p>",
    f"<p>Clientes: {ventas_summary['Clientes_totales']:,}</p>",
    f"<p>Descuentos: {ventas_summary['Descuentos_totales']:,.2f}</p>",
    f"<p>Ticket promedio: {ventas_summary['Ticket_promedio']:,.2f}</p></section>",
    "<section><h2>Riesgos</h2><ul>",
    f"<li>Los descuentos acumulados ascienden a {ventas_summary['Descuentos_totales']:,.2f} y deben revisarse junto con su efecto sobre el margen ponderado de {ventas_summary['Margen_ponderado']:.2%}.</li>",
    f"<li>La región {region_top} lidera las ventas, mientras que la región {region_margen} lidera el margen; esta diferencia requiere seguimiento de rentabilidad por región.</li>",
    f"<li>La concentración de ventas en la categoría {categoria_top} debe vigilarse para evitar depender de un único segmento comercial.</li>",
    "</ul></section>",
    "<section><h2>Oportunidades</h2><ul>",
    f"<li>La categoría {categoria_top} lidera las ventas y la categoría {categoria_margen} lidera el margen, por lo que existe oportunidad de combinar volumen y rentabilidad en la estrategia comercial.</li>",
    f"<li>La región {region_top} ofrece la mayor escala de ventas y la región {region_margen} el mayor margen, lo que permite comparar y replicar prácticas comerciales.</li>",
    f"<li>El producto {producto_top} es el de mayor facturación acumulada y puede priorizarse en el seguimiento de inventario y promoción.</li>",
    "</ul></section>",
    "<section><h2>Conclusiones</h2><ul>",
    f"<li>El producto con mayor facturación acumulada es <strong>{producto_top}</strong>.</li>",
    f"<li>La categoría <strong>{categoria_top}</strong> concentra el mayor volumen de ventas.</li>",
    f"<li>La categoría <strong>{categoria_margen}</strong> presenta el margen más alto.</li>",
    f"<li>La región <strong>{region_top}</strong> lidera las ventas y la región <strong>{region_margen}</strong> lidera el margen.</li>",
    "<li>El resultado debe evaluarse equilibrando volumen, margen y nivel de descuentos.</li>",
    "</ul></section>",
    "<section><h2>Recomendaciones</h2><ul>",
    f"<li>Fortalecer la promoción de <strong>{producto_top}</strong> y medir su rentabilidad, no solo sus ventas.</li>",
    f"<li>Usar la categoría <strong>{categoria_margen}</strong> como referencia para mejorar el margen de las categorías de mayor volumen.</li>",
    f"<li>Conservar la cobertura comercial de <strong>{region_top}</strong> y replicar las prácticas rentables de <strong>{region_margen}</strong>.</li>",
    "<li>Definir límites de descuento y revisar periódicamente su efecto sobre la ganancia.</li>",
    "<li>Actualizar el archivo Excel y ejecutar nuevamente el notebook para comparar periodos.</li>",
    "</ul></section>",
]

for indice, figura in enumerate(figuras_interactivas):
    html_parts.append(figura.to_html(full_html=False, include_plotlyjs='cdn' if indice == 0 else False))

html_parts.append("</body></html>")
html_path = report_dir / 'informe_interactivo.html'
html_path.write_text(''.join(html_parts), encoding='utf-8')
print(f'HTML interactivo creado: {html_path}')

# 2) Generar PDF con resumen, conclusiones, recomendaciones y gráficas
pdf_path = report_dir / 'informe_gerencial.pdf'
styles = getSampleStyleSheet()
doc = SimpleDocTemplate(str(pdf_path), pagesize=letter)
story = [
    Paragraph('Informe gerencial de rentabilidad', styles['Title']),
    Spacer(1, 12),
    Paragraph('Resumen ejecutivo', styles['Heading2']),
    Paragraph(f"Ventas totales: {ventas_summary['Ventas_totales']:,.2f}", styles['BodyText']),
    Paragraph(f"Ganancia total: {ventas_summary['Ganancia_total']:,.2f}", styles['BodyText']),
    Paragraph(f"Margen ponderado: {ventas_summary['Margen_ponderado']:.2%}", styles['BodyText']),
    Paragraph(f"Ticket promedio: {ventas_summary['Ticket_promedio']:,.2f}", styles['BodyText']),
    Spacer(1, 12),
    Paragraph('Conclusiones', styles['Heading2']),
    Paragraph(f'Producto de mayor facturación: {producto_top}. Categoría líder en ventas: {categoria_top}. Categoría con mayor margen: {categoria_margen}. Región líder en ventas: {region_top}.', styles['BodyText']),
    Spacer(1, 12),
    Paragraph('Recomendaciones', styles['Heading2']),
    Paragraph(f'Promover {producto_top}, mejorar el margen de las categorías de alto volumen, replicar las prácticas de {region_margen} y controlar los descuentos.', styles['BodyText']),
    Spacer(1, 20),
]

from reportlab.platypus import PageBreak

for nombre in ['ventas_mensual.png', 'ventas_categoria.png', 'margen_categoria.png', 'ventas_region.png']:
    story.append(PageBreak())
    story.append(Image(str(report_dir / nombre), width=430, height=240))
    story.append(Spacer(1, 12))

doc.build(list(story))
print(f'PDF generado: {pdf_path}')
print('Exportación final completada.')


Note: you may need to restart the kernel to use updated packages.
Directorio de reportes: D:\alejandro_palencia\Curso Python Daxus\Proyecto04\reportes
Imagen creada: reportes\ventas_mensual.png
Imagen creada: reportes\ventas_categoria.png
Imagen creada: reportes\margen_categoria.png
Imagen creada: reportes\ventas_region.png
HTML interactivo creado: reportes\informe_interactivo.html
PDF generado: reportes\informe_gerencial.pdf
Exportación final completada.


In [11]:
# ETAPA 11 — GUARDAR UNA NUEVA COPIA DEL PDF
# Se usa otro nombre por si el PDF anterior está abierto en un visor.
from pathlib import Path
from reportlab.platypus import SimpleDocTemplate

pdf_path_nuevo = Path('reportes') / 'informe_gerencial_1000_registros.pdf'
doc_nuevo = SimpleDocTemplate(str(pdf_path_nuevo), pagesize=letter)
doc_nuevo.build(list(story))
print(f'PDF nuevo generado: {pdf_path_nuevo.resolve()}')


PDF nuevo generado: D:\alejandro_palencia\Curso Python Daxus\Proyecto04\reportes\informe_gerencial_1000_registros.pdf


In [12]:
len(story)

26